Speed up Block placer

jakobwenzel opened the pull request:
This PR (together with other changes that have already landed) improves placer runtime by ~5x, while also allowing for dynamic module implementation selection during placement.

For modules with multiple implementations, the placer used to bind them before placement. This PR introduces ModuleImplsInst as an instance of a module with changeable implementation. BlockPlacer2 has been made generic to allow for it to place either ModuleInsts or ModuleImplsInsts. When using ModuleImplsInsts, this allows the placer to select the instances during placement, leading to better QoR.

The list of runtime-specific improvements is:

- Use `RelocatableTileRectangle` to check for overlap between modules instead of relocating all module `SiteInsts`. After constructing the rectangle, relocation cost is O(1) instead of O(n) in the number of module SiteInsts.
- Cache Net costs as instance of `SimpleTileRectangle`. Only recalculate cost of nets that the module we have moved connects to.
- Optionally ignore Clock Nets. Their cost is slow to calculate, but they end up in global routing. So we don't even gain anything by keeping them as short as possible.
- Region-based overlap checking (class `RegionBasedOverlapCache`). This divides the fabric into square regions of a given side length (default value found by benchmarking). For each region, all modules that touch the region are stored. When a module is moved, overlap detection only needs to be performed for modules that touch the same regions as the module that is being moved.
- Optimized storage of possible module locations (class `SortedValidPlacementCache`). They are stored by position on the fabric together with some additional indices. This allows for optimized querying of placements near a central placement.

eddieh-xlnx commented on src/com/xilinx/rapidwright/design/ModuleImplsInstance.java:
```diff
@@ -0,0 +1,115 @@
+        this(name, null, module);
+    }
+
+    public void unPlace() {
```
Yes I agree, moving `place` would be a bit trickier. Let's skip that for now.

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/BlockPlacer2Module.java:
```diff
@@ -0,0 +1,471 @@
+            SitePinInst src = net.getSource();
+            ArrayList<SitePinInst> snks = new ArrayList<SitePinInst>();
+            if(src == null){
+                // TODO - This should not happen
+                //System.out.println("ERROR: Need to find out why net: " + net.getName() + " has no driver\n\n" + net.toString() );
+                continue;
```
Roger that. Let's leave it for now then.

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/ImplsPath.java:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        int fanOutPenalty = 1;
+        if (getSize() > 30){
+            fanOutPenalty = 3;
```
Righty. Let's skip this one too.

jakobwenzel commented on src/com/xilinx/rapidwright/placer/blockplacer/BlockPlacer2Impls.java:
```diff
@@ -0,0 +1,311 @@
+            if (cell != null) {
+                List<String> siteWires = new ArrayList<>();
+                SitePinInst spi = cell.getSitePinFromPortInst(portInst, siteWires);
+                if (spi == null) {
+                    System.out.println(siteWires);
+                    System.out.println(cell.getAllSitePinsFromPortInst(portInst, siteWires));
+                    System.out.println(siteWires);
+                    System.out.println(cell.getSitePinFromPortInst(portInst, siteWires));
+                    System.out.println(siteWires);
+                    //TODO is this allowed to happen?
+                    System.err.println("while creating an ImplsInstancePort for "+portInst+", could not find the port in cell "+cell);
+                    return null;
+                }
```
I think I added this while debugging a design where physical net names were not consistent. Users should call `DesignTools.makePhysNetNamesConsistent` before running the placer. I'll change this to throw an exception pointing this out.

jakobwenzel commented on src/com/xilinx/rapidwright/placer/blockplacer/OverlapCache.java:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * Magic Size found by benchmarking
+     */
+    public static int DEFAULT_SIZE = 23;
```
It's used in `BlockPlacer2Impls`' constructor: https://github.com/Xilinx/RapidWright/blob/277e9e82c23671e987e6acc1177adb4ac54caf2f/src/com/xilinx/rapidwright/placer/blockplacer/BlockPlacer2Impls.java#L69-L71

eddieh-xlnx replied:
Oooh right. Perhaps this could be made clearer by creating a `OverlapCache` that uses the value locally?

jakobwenzel commented:
I think I have addressed all review comments. I've also added the testcase. Since I've introduced more command line switches in PicoBlazeArray's main, I am now using a real command line parser there. Checkpoint Directory, Part and output filename now have to be prefixed by an appropriate switch (e.g. --part). Is this change ok or should I revert that? The [tutorial](https://www.rapidwright.io/docs/PreImplemented_Modules_Part_II.html) would have to be changed accordingly.


I also removed the Bounding Box implementation inside `HardMacro`. When modules are moved to straddle a clock distribution row/column, their size changes. That implementation did not account for that, while `RelocatableTileRectangle` does.


The testcase depends on https://github.com/eddieh-xlnx/RapidWrightDCP/pull/2 but CI ran through anyway. It seems we can point the submodule commit at something that is not yet merged and should therefore only exist inside my fork...

eddieh-xlnx commented:
> The testcase depends on eddieh-xlnx/RapidWrightDCP#2 but CI ran through anyway. It seems we can point the submodule commit at something that is not yet merged and should therefore only exist inside my fork...

I think this is some GitHub magic that makes commits from downstream forks available to upstream. More evidence of this is that your commit https://github.com/jakobwenzel/RapidWrightDCP/commit/3b1373f44342c02cdea8ea36efb6f55c158a600a is also available at https://github.com/eddieh-xlnx/RapidWrightDCP/commit/3b1373f44342c02cdea8ea36efb6f55c158a600a but with the warning:
> This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.

Neat!

eddieh-xlnx commented on src/com/xilinx/rapidwright/design/AbstractModuleInst.java:
```diff
@@ -0,0 +1,209 @@
+import com.xilinx.rapidwright.edif.EDIFPort;
+
+/**
+ * Abstract base class for {@link ModuleInst} and {@link ModuleImplsInstance}/{@link ModuleImplsInst}
```
Removing the pre-rename suggestion.
```suggestion
 * Abstract base class for {@link ModuleInst} and {@link ModuleImplsInst}
```

eddieh-xlnx commented on src/com/xilinx/rapidwright/design/DesignTools.java:
```diff
@@ -2539,4 +2548,97 @@ public static Path generateReadableEDIF(Path dcp, Path edfFileName) {
+	 * @param module the module to use
+	 * @return the newly created instance
+	 */
+	public static ModuleImplsInst createModuleImplsInstance(Design design, String name, ModuleImpls module) {
```
```suggestion
	public static ModuleImplsInst createModuleImplsInst(Design design, String name, ModuleImpls module) {
```

eddieh-xlnx commented on src/com/xilinx/rapidwright/examples/PicoBlazeArray.java:
```diff
@@ -54,122 +76,279 @@ public static void updateAnchorToBRAM(Module m){
+
+			@Override
+			protected ModuleImplsInst createInstance(Design design, String name, Module impl, ModuleImpls impls) {
+				return DesignTools.createModuleImplsInstance(design, name, impls);
```
```suggestion
				return DesignTools.createModuleImplsInst(design, name, impls);
```

eddieh-xlnx commented on src/com/xilinx/rapidwright/gui/UiPBlock.java:
```diff
@@ -0,0 +1,48 @@
+import com.xilinx.rapidwright.design.TileRectangle;
+
+/**
+ * A named and colored PBlock that can be shown in a {@link PblockScene}
```
```suggestion
 * A named and colored PBlock that can be shown in a {@link PBlockScene}
```

eddieh-xlnx commented on src/com/xilinx/rapidwright/gui/PblockScene.java:
```diff
@@ -0,0 +1,175 @@
+/**
+ * UI Scene that can show named PBlocks
+ */
+public class PblockScene extends TileScene {
```
Rename `PBlockScene` for consistency.

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/OverlapCache.java:
```diff
@@ -0,0 +1,212 @@
+        allTouchedRegionsMatch(mii,l->{l.add(mii); return true;});
+    }
+
+    public OverlapCache(Device device, Collection<ModuleImplsInst> instances, int size) {
```
`size` -> `regionSize`

eddieh-xlnx reviewed (commented):
Thanks for resolving all my comments!! Just a handful of extra things I noticed on my second pass.

> Since I've introduced more command line switches in PicoBlazeArray's main, I am now using a real command line parser there. Checkpoint Directory, Part and output filename now have to be prefixed by an appropriate switch (e.g. --part). Is this change ok or should I revert that? The tutorial would have to be changed accordingly.

I would suggest we revert. I've left a suggestion in PicoBlazeArray.

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/OverlapCache.java:
```diff
@@ -0,0 +1,212 @@
+ * the region are stored. When a module is moved, overlap detection only needs to be performed for modules that
+ * touch the same regions as the module that is being moved.
+ */
+public class OverlapCache extends AbstractOverlapCache {
```
(Don't forget to update PR description too)

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/AbstractPlacementCollection.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Manages possible placements of a single module. This does not take overlaps with other modules into account. 
+ */
+public abstract class AbstractPlacementCollection<PlacementT> {
```
Would `AbstractValidPlacementCache` be a better name?

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/DumbPlacementCollection.java:
```diff
@@ -0,0 +1,61 @@
+ * Naive implementation of placement collection
+ * @param <PlacementT> The placement class
+ */
+public class DumbPlacementCollection<PlacementT> extends AbstractPlacementCollection<PlacementT>{
```
`ExhaustiveValidPlacementCache`?

eddieh-xlnx commented on src/com/xilinx/rapidwright/placer/blockplacer/PlacementCollection.java:
```diff
@@ -0,0 +1,189 @@
+ * This allows for optimized querying of placements near a central placement.
+ * @param <PlacementT> The placement class
+ */
+public class PlacementCollection<PlacementT> extends AbstractPlacementCollection<PlacementT> {
```
`SortedValidPlacementCache`? (also update PR desc)

jakobwenzel commented on src/com/xilinx/rapidwright/placer/blockplacer/BlockPlacer2Impls.java:
```diff
@@ -0,0 +1,295 @@
+        super(design, ignoreMostUsedNets, graphData);
+
+        this.moduleInstances = new ArrayList<>(moduleInstances);
+        overlaps = new OverlapCache(design.getDevice(), moduleInstances, overlapSize);
```
Good idea. This also makes usage of OverlapCache.DEFAULT_SIZE local to that class.

jakobwenzel commented on src/com/xilinx/rapidwright/examples/PicoBlazeArray.java:
```diff
@@ -54,122 +76,279 @@ public static void updateAnchorToBRAM(Module m){
+			}
+			throw e;
+		}
+
```
Oohh, nice. I totally missed that method in their docs.

I've changed the code to use this. Before any of my changes, --no-hand-placer only worked as last argument. All named options now work everywhere in the command line. 

However, there is one regression: When one of the non-option arguments starts with "-", a very specific order now has to be followed: Named options have to be given first, then "--", then the non-option arguments.

I think we can live with that.

eddieh-xlnx replied:
> When one of the non-option arguments starts with "-"

Great spot. I think if you have a file name starting with "-" then you're asking for trouble!

jakobwenzel commented:
I've addressed all comments from the second round of review.

The submodule is still pointing into my fork for now.

eddieh-xlnx commented:
> 
> 
> I've addressed all comments from the second round of review.
> 
> The submodule is still pointing into my fork for now.

Looks great, thanks for all your patience!. I've just merge eddieh-xlnx/RapidWrightDCP#2 now, so please update the submodule and then this will be good to go!

I plan on adding a GH check to ensure that we don't ever commit submodule references pointing to downstream forks too.

jakobwenzel commented:
> Looks great, thanks for all your patience!
Thanks for the review, I think your comments really improved my code.

>I've just merge [eddieh-xlnx/RapidWrightDCP#2](https://github.com/eddieh-xlnx/RapidWrightDCP/pull/2) now, so please update the submodule and then this will be good to go!
I just pushed a commit doing just that.

> I plan on adding a GH check to ensure that we don't ever commit submodule references pointing to downstream forks too.
Sounds good!

eddieh-xlnx reviewed (approved):
All good! I don't have the power to resolve some of the last few conversations. Would you mind clearing those out? Thanks!

jakobwenzel commented:
I pressed all the resolve buttons I was able to find.
